File: src/storage/repository.py
"""数据访问层 —— 封装 CRUD 操作。"""

import datetime
import hashlib
import json
from typing import Optional

from sqlalchemy import func, desc
from sqlalchemy.orm import Session

from .models import (
    KnowledgeDomain, KnowledgeItem, PushHistory,
    UserSettings, ChatMessage, AgentExecutionLog
)


# ── 领域 CRUD ──

def get_all_domains(session: Session) -> list[KnowledgeDomain]:
    return session.query(KnowledgeDomain).order_by(KnowledgeDomain.sort_order).all()


def get_enabled_domains(session: Session) -> list[KnowledgeDomain]:
    return (
        session.query(KnowledgeDomain)
        .filter(KnowledgeDomain.is_enabled == True)
        .order_by(KnowledgeDomain.sort_order)
        .all()
    )


def get_domain_by_id(session: Session, domain_id: int) -> Optional[KnowledgeDomain]:
    return session.query(KnowledgeDomain).filter(KnowledgeDomain.id == domain_id).first()


def create_domain(session: Session, name: str, description: str = "", keywords: str = "", icon: str = "📚") -> KnowledgeDomain:
    max_order = session.query(func.max(KnowledgeDomain.sort_order)).scalar() or 0
    domain = KnowledgeDomain(
        name=name, description=description, keywords=keywords,
        icon=icon, sort_order=max_order + 1
    )
    session.add(domain)
    session.commit()
    return domain


def update_domain(session: Session, domain_id: int, **kwargs) -> Optional[KnowledgeDomain]:
    domain = get_domain_by_id(session, domain_id)
    if not domain:
        return None
    for k, v in kwargs.items():
        if hasattr(domain, k):
            setattr(domain, k, v)
    domain.updated_at = datetime.datetime.utcnow()
    session.commit()
    return domain


def delete_domain(session: Session, domain_id: int) -> bool:
    domain = get_domain_by_id(session, domain_id)
    if not domain:
        return False
    session.delete(domain)
    session.commit()
    return True


# ── 知识条目 CRUD ──

def create_knowledge_item(session: Session, domain_id: int, domain_name: str,
                          title: str, summary: str, detail: str,
                          source_url: str = None, source_title: str = None,
                          trust_score: float = 0.5) -> Optional[KnowledgeItem]:
    raw = f"{title}|{summary}|{detail}"
    content_hash = hashlib.sha256(raw.encode()).hexdigest()

    existing = session.query(KnowledgeItem).filter(KnowledgeItem.content_hash == content_hash).first()
    if existing:
        return None

    item = KnowledgeItem(
        domain_id=domain_id,
        domain_name=domain_name,
        title=title,
        summary=summary,
        detail=detail,
        source_url=source_url,
        source_title=source_title,
        trust_score=trust_score,
        content_hash=content_hash,
    )
    session.add(item)
    session.commit()
    return item


def get_knowledge_items(session: Session, domain_id: int = None,
                        keyword: str = None, limit: int = 50, offset: int = 0) -> list[KnowledgeItem]:
    q = session.query(KnowledgeItem)
    if domain_id:
        q = q.filter(KnowledgeItem.domain_id == domain_id)
    if keyword:
        like = f"%{keyword}%"
        q = q.filter(
            KnowledgeItem.title.like(like) | KnowledgeItem.summary.like(like)
        )
    return q.order_by(desc(KnowledgeItem.created_at)).offset(offset).limit(limit).all()


def get_knowledge_item_by_id(session: Session, item_id: int) -> Optional[KnowledgeItem]:
    return session.query(KnowledgeItem).filter(KnowledgeItem.id == item_id).first()


def update_knowledge_item(session: Session, item_id: int, **kwargs) -> Optional[KnowledgeItem]:
    item = get_knowledge_item_by_id(session, item_id)
    if not item:
        return None
    for k, v in kwargs.items():
        if hasattr(item, k):
            setattr(item, k, v)
    session.commit()
    return item


def get_knowledge_count(session: Session) -> int:
    return session.query(func.count(KnowledgeItem.id)).scalar() or 0


def get_recent_count(session: Session, days: int = 7) -> int:
    since = datetime.datetime.utcnow() - datetime.timedelta(days=days)
    return (
        session.query(func.count(KnowledgeItem.id))
        .filter(KnowledgeItem.created_at >= since)
        .scalar() or 0
    )


# ── 推送历史 ──

def create_push_history(session: Session, item_id: int) -> PushHistory:
    entry = PushHistory(knowledge_item_id=item_id)
    session.add(entry)
    session.commit()
    return entry


def get_recent_push_history(session: Session, limit: int = 20) -> list[PushHistory]:
    return (
        session.query(PushHistory)
        .order_by(desc(PushHistory.pushed_at))
        .limit(limit)
        .all()
    )


# ── 用户设置 ──

def get_setting(session: Session, key: str) -> Optional[str]:
    row = session.query(UserSettings).filter(UserSettings.key == key).first()
    return row.value if row else None


def set_setting(session: Session, key: str, value: str):
    row = session.query(UserSettings).filter(UserSettings.key == key).first()
    if row:
        row.value = value
    else:
        row = UserSettings(key=key, value=value)
        session.add(row)
    session.commit()


def get_all_settings(session: Session) -> dict:
    rows = session.query(UserSettings).all()
    result = {}
    for row in rows:
        try:
            result[row.key] = json.loads(row.value)
        except (json.JSONDecodeError, TypeError):
            result[row.key] = row.value
    return result


# ── 对话消息 ──

def save_chat_message(session: Session, session_id: str, role: str, content: str) -> ChatMessage:
    msg = ChatMessage(session_id=session_id, role=role, content=content)
    session.add(msg)
    session.commit()
    return msg


def get_chat_history(session: Session, session_id: str, limit: int = 50) -> list[ChatMessage]:
    return (
        session.query(ChatMessage)
        .filter(ChatMessage.session_id == session_id)
        .order_by(ChatMessage.created_at)
        .limit(limit)
        .all()
    )


# ── Agent 执行日志 ──

def save_agent_log(session: Session, session_id: str, step_type: str,
                   tool_name: str = None, tool_input: str = None,
                   tool_output: str = None, content: str = None) -> AgentExecutionLog:
    log = AgentExecutionLog(
        session_id=session_id,
        step_type=step_type,
        tool_name=tool_name,
        tool_input=tool_input,
        tool_output=tool_output,
        content=content,
    )
    session.add(log)
    session.commit()
    return log


# ── 领域统计 (用于 Agent) ──

def get_domain_stats(session: Session) -> list[dict]:
    rows = (
        session.query(
            KnowledgeItem.domain_id,
            KnowledgeItem.domain_name,
            func.count(KnowledgeItem.id).label("count"),
            func.avg(KnowledgeItem.rating).label("avg_rating"),
        )
        .group_by(KnowledgeItem.domain_id, KnowledgeItem.domain_name)
        .all()
    )
    return [
        {"domain_id": r.domain_id, "domain_name": r.domain_name,
         "count": r.count, "avg_rating": round(r.avg_rating or 0, 1)}
        for r in rows
    ]


def get_user_feedback(session: Session, limit: int = 30) -> list[dict]:
    items = (
        session.query(KnowledgeItem)
        .filter(KnowledgeItem.rating != None)
        .order_by(desc(KnowledgeItem.created_at))
        .limit(limit)
        .all()
    )
    return [
        {"domain_name": i.domain_name, "title": i.title,
         "rating": i.rating, "is_favorited": i.is_favorited}
        for i in items
    ]

File: desktop/src/core/repository.ts
import { getDb } from "./db";
import { DEFAULT_SETTINGS } from "./defaults";
import type { AgentLog, AppSettings, DashboardStats, Domain, DomainStat, KnowledgeItem } from "./types";

export async function listDomains(): Promise<Domain[]> {
  const db = await getDb();
  return db.select<Domain[]>(
    "SELECT id, name, description, keywords, sort_order, is_enabled FROM knowledge_domains ORDER BY sort_order ASC, id ASC"
  );
}

export async function saveDomain(input: Partial<Domain> & Pick<Domain, "name">): Promise<void> {
  const db = await getDb();
  if (input.id) {
    await db.execute(
      "UPDATE knowledge_domains SET name = ?, description = ?, keywords = ?, is_enabled = ? WHERE id = ?",
      [input.name, input.description ?? "", input.keywords ?? "", input.is_enabled ?? 1, input.id]
    );
    return;
  }
  await db.execute(
    "INSERT INTO knowledge_domains (name, description, keywords, sort_order, is_enabled) VALUES (?, ?, ?, COALESCE((SELECT MAX(sort_order) + 1 FROM knowledge_domains), 0), ?)",
    [input.name, input.description ?? "", input.keywords ?? "", input.is_enabled ?? 1]
  );
}

export async function deleteDomain(domainId: number): Promise<void> {
  const db = await getDb();
  await db.execute("DELETE FROM knowledge_domains WHERE id = ?", [domainId]);
}

export async function listKnowledge(keyword = ""): Promise<KnowledgeItem[]> {
  const db = await getDb();
  if (!keyword.trim()) {
    return db.select<KnowledgeItem[]>(
      "SELECT * FROM knowledge_items ORDER BY datetime(created_at) DESC LIMIT 100"
    );
  }
  const like = `%${keyword.trim()}%`;
  return db.select<KnowledgeItem[]>(
    "SELECT * FROM knowledge_items WHERE title LIKE ? OR summary LIKE ? OR detail LIKE ? ORDER BY datetime(created_at) DESC LIMIT 100",
    [like, like, like]
  );
}

export async function getKnowledgeById(itemId: number): Promise<KnowledgeItem | null> {
  const db = await getDb();
  const rows = await db.select<KnowledgeItem[]>(
    "SELECT * FROM knowledge_items WHERE id = ? LIMIT 1",
    [itemId]
  );
  return rows[0] ?? null;
}

export async function deleteKnowledgeItem(itemId: number): Promise<void> {
  const db = await getDb();
  await db.execute("DELETE FROM push_history WHERE knowledge_item_id = ?", [itemId]);
  await db.execute("DELETE FROM knowledge_items WHERE id = ?", [itemId]);
}

export async function clearAgentLogs(): Promise<void> {
  const db = await getDb();
  await db.execute("DELETE FROM agent_execution_logs");
}

export async function createKnowledgeCard(input: {
  domainId: number | null;
  domainName: string;
  title: string;
  summary: string;
  detail: string;
  sourceUrl?: string;
  sourceTitle?: string;
}): Promise<{ status: "created"; itemId: number } | { status: "duplicate" }> {
  const db = await getDb();
  const hash = await sha256(`${input.title}|${input.summary}|${input.detail}`);
  const exists = await db.select<{ id: number }[]>(
    "SELECT id FROM knowledge_items WHERE content_hash = ? LIMIT 1",
    [hash]
  );
  if (exists.length > 0) {
    return { status: "duplicate" };
  }
  await db.execute(
    "INSERT INTO knowledge_items (domain_id, domain_name, title, summary, detail, source_url, source_title, trust_score, content_hash) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
    [
      input.domainId,
      input.domainName,
      input.title,
      input.summary,
      input.detail,
      input.sourceUrl ?? null,
      input.sourceTitle ?? null,
      0.5,
      hash
    ]
  );
  const item = await db.select<{ id: number }[]>("SELECT last_insert_rowid() as id");
  const itemId = item[0]?.id;
  if (itemId) {
    await db.execute("INSERT INTO push_history (knowledge_item_id) VALUES (?)", [itemId]);
    return { status: "created", itemId };
  }
  return { status: "duplicate" };
}

export async function getSettings(): Promise<AppSettings> {
  const db = await getDb();
  const rows = await db.select<{ key: string; value: string }[]>(
    "SELECT key, value FROM user_settings"
  );
  const map: Record<string, unknown> = { ...DEFAULT_SETTINGS };
  for (const row of rows) {
    try {
      map[row.key] = JSON.parse(row.value);
    } catch {
      map[row.key] = row.value;
    }
  }
  return map as unknown as AppSettings;
}

export async function saveSettings(settings: AppSettings): Promise<void> {
  const db = await getDb();
  for (const [key, value] of Object.entries(settings)) {
    await db.execute(
      "INSERT INTO user_settings (key, value) VALUES (?, ?) ON CONFLICT(key) DO UPDATE SET value=excluded.value",
      [key, JSON.stringify(value)]
    );
  }
}

export async function getDashboardStats(): Promise<DashboardStats> {
  const db = await getDb();
  const total = await db.select<{ count: number }[]>("SELECT COUNT(*) as count FROM knowledge_items");
  const recent7d = await db.select<{ count: number }[]>(
    "SELECT COUNT(*) as count FROM knowledge_items WHERE datetime(created_at) >= datetime('now', '-7 day')"
  );
  const domains = await db.select<{ count: number }[]>(
    "SELECT COUNT(*) as count FROM knowledge_domains WHERE is_enabled = 1"
  );
  return {
    total: total[0]?.count ?? 0,
    recent7d: recent7d[0]?.count ?? 0,
    domains: domains[0]?.count ?? 0
  };
}

export async function saveAgentLog(input: {
  sessionId: string;
  stepType: AgentLog["step_type"];
  content?: string;
  toolName?: string;
  toolInput?: string;
  toolOutput?: string;
}): Promise<void> {
  const db = await getDb();
  await db.execute(
    "INSERT INTO agent_execution_logs (session_id, step_type, tool_name, tool_input, tool_output, content) VALUES (?, ?, ?, ?, ?, ?)",
    [
      input.sessionId,
      input.stepType,
      input.toolName ?? null,
      input.toolInput ?? null,
      input.toolOutput ?? null,
      input.content ?? null
    ]
  );
}

export async function getRecentAgentLogs(limit = 50): Promise<AgentLog[]> {
  const db = await getDb();
  return db.select<AgentLog[]>(
    "SELECT * FROM agent_execution_logs ORDER BY id DESC LIMIT ?",
    [limit]
  );
}

export async function getRecentPushHistory(
  limit = 8
): Promise<Array<{ id: number; knowledge_item_id: number | null; pushed_at: string; title: string; domain_name: string }>> {
  const db = await getDb();
  return db.select<Array<{ id: number; knowledge_item_id: number | null; pushed_at: string; title: string; domain_name: string }>>(
    `SELECT ph.id, ph.knowledge_item_id, ph.pushed_at, COALESCE(ki.title, '(已删除)') as title, COALESCE(ki.domain_name, '') as domain_name
     FROM push_history ph
     LEFT JOIN knowledge_items ki ON ki.id = ph.knowledge_item_id
     ORDER BY ph.id DESC
     LIMIT ?`,
    [limit]
  );
}

export async function getDomainStats(): Promise<DomainStat[]> {
  const db = await getDb();
  return db.select<DomainStat[]>(
    `SELECT
      domain_id,
      COALESCE(domain_name, '') as domain_name,
      COUNT(*) as count,
      COALESCE(ROUND(AVG(rating), 1), 0) as avg_rating
    FROM knowledge_items
    GROUP BY domain_id, domain_name
    ORDER BY count DESC`
  );
}

export async function getUserFeedback(
  limit = 20
): Promise<Array<{ domain_name: string; title: string; rating: number | null; is_favorited: number }>> {
  const db = await getDb();
  return db.select<Array<{ domain_name: string; title: string; rating: number | null; is_favorited: number }>>(
    `SELECT
      COALESCE(domain_name, '') as domain_name,
      title,
      rating,
      is_favorited
    FROM knowledge_items
    WHERE rating IS NOT NULL
    ORDER BY datetime(created_at) DESC
    LIMIT ?`,
    [limit]
  );
}

async function sha256(raw: string): Promise<string> {
  const data = new TextEncoder().encode(raw);
  const digest = await crypto.subtle.digest("SHA-256", data);
  return Array.from(new Uint8Array(digest))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("");
}
